osf/utils/machines.py

```diff
@@ -344,118 +345,105 @@ def get_context(self):
-        user = ev.kwargs.get('user')
-        self.machineable.refresh_from_db()
-        self.machineable.registered_node.retract_registration(user)
-
-    def request_terminate_embargo(self, ev):
-        user = ev.kwargs.get('user')
-        self.machineable.refresh_from_db()
-
-        approval = self.machineable.registered_node.request_embargo_termination(user)
-        approval.save()
-
-    def withdraw_registration(self, ev):
-        user = ev.kwargs.get('user')
-        self.machineable.registered_node.retraction._on_complete(user)
+class SanctionStateMachine(Machine):
+    '''SanctionsStateMachine manages state transitions for Sanctions objects.
 
-    def reject_withdrawal(self, ev):
-        user = ev.kwargs.get('user')
-        self.machineable.registered_node.retraction._on_reject(user)
+    The valid machine states for a Sanction object are defined in Workflows.SanctionStates.
+    The valid transitions between these states are defined in Workflows.SANCTION_TRANSITIONS.
 
-    def embargo_registration(self, ev):
-        user = ev.kwargs.get('user')
-        end_date = ev.kwargs.get('end_date')
-        self.machineable.refresh_from_db()
-        self.machineable.registered_node.embargo_registration(user, end_date)
-        self.machineable.registered_node.refresh_from_db()
+    Subclasses of SanctionStateMachine inherit the 'trigger' functions named in
+    the SANCTION_TRANSITIONS dictionary (approve, accept, and reject).
+    These trigger functions will, in order,
+    1) Call any 'prepare_event' functions defined on the StateMachine (see __init__)
+    2) Call Sanction member functions listed in the 'conditions' key of the dictionary
+    3) Call Sanction member functions listed in the 'before' key of the dictionary
+    4) Update the state field of the Sanction object via the approval_stage setter
+    5) Call Sanction member functions listed in the 'after' key of the dictionary
 
-    def terminate_embargo(self, ev):
-        self.machineable.registered_node.terminate_embargo()
+    If any step fails, the whole transition will fail and the Sanction's
+    approval_stage will be rolled back.
 
-    def edit_comment(self, ev):
-        self.machineable.comment = self.action.comment
-        self.machineable.save()
+    SanctionStateMachine also provides some extra functionality to write
+    RegistrationActions on events moving in to or out of Moderated machine states
+    as well as to convert MachineErrors (which arise on unsupported state changes
+    requests) into HTTPErrors to report back to users who try to initiate such errors.
+    '''
 
-    def accept_draft_registration(self, ev):
-        approval = self.machineable.approval
-        approval.approve(self.action.creator)
-        approval.save()
+    SANCTION_TYPE = SanctionTypes.UNDEFINED
+    MACHINE_STATE_FIELD_NAME = ''
 
-    def reject_draft_registration(self, ev):
-        self.machineable.meta = {}
-        self.machineable.save()
+    def __init__(self):
```

Comment:
I noticed while working on the notifications that overriding this was causing a lot of test fails where we were creating Sanctions.

Might want to try something like: 
```python
@classmethod
def create(cls, *args, **kwargs):
    sanction = cls(*args, **kwargs)
    sanction.initalize_machine()
    return sanction
```

Then put the machine class init in that because here you're overriding the model.